File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-csharp/src/SDKWork.Mail.Sdk/mailProviderPackageLoader.cs
namespace Sdkwork.Mail.Sdk;

using System;
using System.Collections.Generic;
using System.Linq;

public sealed class MailProviderPackageLoaderException : Exception
{
    public MailProviderPackageLoaderException(string code, string message)
        : base(message)
    {
        this.code = code;
    }

    public string code { get; }
}

public sealed record MailProviderPackageLoadRequest(
    string? providerKey = null,
    string? packageIdentity = null
);

public sealed record MailResolvedProviderPackageLoadTarget(
    MailProviderPackageCatalogEntry packageEntry
);

public delegate object? MailProviderPackageImportFn(MailResolvedProviderPackageLoadTarget target);
public delegate object? MailProviderPackageLoaderFn(MailProviderPackageLoadRequest request);

public sealed record MailProviderPackageInstallRequest(
    object driverManager,
    MailProviderPackageLoadRequest loadRequest
);

public static class MailProviderPackageLoader
{
    public const string ProviderPackageNotFound = "provider_package_not_found";
    public const string ProviderPackageIdentityMismatch = "provider_package_identity_mismatch";
    public const string ProviderPackageLoadFailed = "provider_package_load_failed";
    public const string ProviderModuleExportMissing = "provider_module_export_missing";

    public static MailResolvedProviderPackageLoadTarget ResolveMailProviderPackageLoadTarget(
        MailProviderPackageLoadRequest? request
    )
    {
        request ??= new MailProviderPackageLoadRequest();
        var packageByProviderKey = string.IsNullOrWhiteSpace(request.providerKey)
            ? null
            : MailProviderPackageCatalog.GetMailProviderPackageByProviderKey(request.providerKey!);
        var packageByIdentity = string.IsNullOrWhiteSpace(request.packageIdentity)
            ? null
            : MailProviderPackageCatalog.GetMailProviderPackageByPackageIdentity(request.packageIdentity!);

        if (packageByProviderKey is not null
            && packageByIdentity is not null
            && !string.Equals(packageByProviderKey.packageIdentity, packageByIdentity.packageIdentity, StringComparison.Ordinal))
        {
            throw new MailProviderPackageLoaderException(
                ProviderPackageIdentityMismatch,
                "providerKey and packageIdentity must resolve to the same provider package boundary."
            );
        }

        var resolvedPackage = packageByProviderKey ?? packageByIdentity;
        if (resolvedPackage is null)
        {
            throw new MailProviderPackageLoaderException(
                ProviderPackageNotFound,
                "No official provider package matches the requested provider boundary."
            );
        }

        return new MailResolvedProviderPackageLoadTarget(resolvedPackage);
    }

    public static MailProviderPackageLoaderFn CreateMailProviderPackageLoader(
        MailProviderPackageImportFn importPackage
    ) => request => LoadMailProviderModule(request, importPackage);

    public static object? LoadMailProviderModule(
        MailProviderPackageLoadRequest request,
        MailProviderPackageImportFn importPackage
    )
    {
        var target = ResolveMailProviderPackageLoadTarget(request);

        try
        {
            var providerModule = importPackage(target);
            if (providerModule is null)
            {
                throw new MailProviderPackageLoaderException(
                    ProviderModuleExportMissing,
                    "Reserved provider package loader scaffold requires an executable provider module namespace."
                );
            }

            return providerModule;
        }
        catch (MailProviderPackageLoaderException)
        {
            throw;
        }
        catch (Exception error)
        {
            throw new MailProviderPackageLoaderException(
                ProviderPackageLoadFailed,
                $"Reserved provider package loader scaffold could not load {target.packageEntry.packageIdentity}: {error.Message}"
            );
        }
    }

    public static void InstallMailProviderPackage(
        MailProviderPackageInstallRequest request,
        MailProviderPackageImportFn importPackage
    )
    {
        _ = LoadMailProviderModule(request.loadRequest, importPackage);

        throw new MailProviderPackageLoaderException(
            ProviderPackageLoadFailed,
            "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands."
        );
    }

    public static void InstallMailProviderPackages(
        IReadOnlyList<MailProviderPackageInstallRequest> requests,
        MailProviderPackageImportFn importPackage
    )
    {
        foreach (var request in requests)
        {
            _ = LoadMailProviderModule(request.loadRequest, importPackage);
        }

        if (requests.Count > 0)
        {
            throw new MailProviderPackageLoaderException(
                ProviderPackageLoadFailed,
                "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands."
            );
        }
    }
}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-java/src/main/java/com/sdkwork/Mail/metadata/mailProviderPackageLoader.java
package com.sdkwork.Mail.metadata;

import java.util.List;
import java.util.Optional;

public final class MailProviderPackageLoader {

  public static final String PROVIDER_PACKAGE_NOT_FOUND = "provider_package_not_found";
  public static final String PROVIDER_PACKAGE_IDENTITY_MISMATCH = "provider_package_identity_mismatch";
  public static final String PROVIDER_PACKAGE_LOAD_FAILED = "provider_package_load_failed";
  public static final String PROVIDER_MODULE_EXPORT_MISSING = "provider_module_export_missing";

  public record MailProviderPackageLoadRequest(
      String providerKey,
      String packageIdentity
  ) {
  }

  public record MailResolvedProviderPackageLoadTarget(
      MailProviderPackageCatalog.MailProviderPackageCatalogEntry packageEntry
  ) {
  }

  @FunctionalInterface
  public interface MailProviderPackageImportFn {
    Object importPackage(MailResolvedProviderPackageLoadTarget target);
  }

  @FunctionalInterface
  public interface MailProviderPackageLoaderFn {
    Object load(MailProviderPackageLoadRequest request);
  }

  public record MailProviderPackageInstallRequest(
      Object driverManager,
      MailProviderPackageLoadRequest loadRequest
  ) {
  }

  public static MailResolvedProviderPackageLoadTarget resolveMailProviderPackageLoadTarget(
      MailProviderPackageLoadRequest request
  ) {
    var resolvedRequest = request == null
        ? new MailProviderPackageLoadRequest(null, null)
        : request;
    var packageByProviderKey = Optional.ofNullable(resolvedRequest.providerKey())
        .flatMap(MailProviderPackageCatalog::getMailProviderPackageByProviderKey);
    var packageByIdentity = Optional.ofNullable(resolvedRequest.packageIdentity())
        .flatMap(MailProviderPackageCatalog::getMailProviderPackageByPackageIdentity);

    if (packageByProviderKey.isPresent() && packageByIdentity.isPresent()
        && !packageByProviderKey.get().packageIdentity().equals(packageByIdentity.get().packageIdentity())) {
      throw new MailProviderPackageLoaderException(
          PROVIDER_PACKAGE_IDENTITY_MISMATCH,
          "providerKey and packageIdentity must resolve to the same provider package boundary."
      );
    }

    var resolvedPackage = packageByProviderKey.or(() -> packageByIdentity).orElseThrow(() ->
        new MailProviderPackageLoaderException(
            PROVIDER_PACKAGE_NOT_FOUND,
            "No official provider package matches the requested provider boundary."
        )
    );

    return new MailResolvedProviderPackageLoadTarget(resolvedPackage);
  }

  public static MailProviderPackageLoaderFn createMailProviderPackageLoader(
      MailProviderPackageImportFn importPackage
  ) {
    return request -> loadMailProviderModule(request, importPackage);
  }

  public static Object loadMailProviderModule(
      MailProviderPackageLoadRequest request,
      MailProviderPackageImportFn importPackage
  ) {
    var target = resolveMailProviderPackageLoadTarget(request);

    try {
      var namespace = importPackage.importPackage(target);
      if (namespace == null) {
        throw new MailProviderPackageLoaderException(
            PROVIDER_MODULE_EXPORT_MISSING,
            "Reserved provider package loader scaffold requires an executable provider module namespace."
        );
      }

      return namespace;
    } catch (MailProviderPackageLoaderException error) {
      throw error;
    } catch (RuntimeException error) {
      throw new MailProviderPackageLoaderException(
          PROVIDER_PACKAGE_LOAD_FAILED,
          "Reserved provider package loader scaffold could not load "
              + target.packageEntry().packageIdentity()
              + ": "
              + error.getMessage()
      );
    }
  }

  public static void installMailProviderPackage(
      MailProviderPackageInstallRequest request,
      MailProviderPackageImportFn importPackage
  ) {
    loadMailProviderModule(request.loadRequest(), importPackage);

    throw new MailProviderPackageLoaderException(
        PROVIDER_PACKAGE_LOAD_FAILED,
        "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands."
    );
  }

  public static void installMailProviderPackages(
      List<MailProviderPackageInstallRequest> requests,
      MailProviderPackageImportFn importPackage
  ) {
    for (var request : requests) {
      loadMailProviderModule(request.loadRequest(), importPackage);
    }

    if (!requests.isEmpty()) {
      throw new MailProviderPackageLoaderException(
          PROVIDER_PACKAGE_LOAD_FAILED,
          "Reserved provider package installer scaffold cannot register provider modules until a verified runtime bridge lands."
      );
    }
  }

  public static final class MailProviderPackageLoaderException extends RuntimeException {
    private final String code;

    public MailProviderPackageLoaderException(String code, String message) {
      super(message);
      this.code = code;
    }

    public String code() {
      return code;
    }
  }

  private MailProviderPackageLoader() {
  }
}
